cpp/src/io/csv/csv_writer.cu

```diff
@@ -0,0 +1,284 @@
+            case GDF_INT64:
+            case GDF_FLOAT32:
+            case GDF_FLOAT64:
+            case GDF_DATE64:
+                if( (unsigned int)col->size != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << (int)col->size << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            case GDF_STRING:
+                strs = (NVStrings*)col->data;
+                if( strs->size() != rows )
+                {
+                    std::cerr << "write_csv: columns sizes do not match (" << strs->size() << "!=" << rows << ")\n";
+                    return GDF_COLUMN_SIZE_MISMATCH;
+                }
+                break;
+            default:
+                std::cerr << "write_csv: unknown column type: " << (int)col->dtype << "\n";
+                return GDF_UNSUPPORTED_DTYPE;
+        }
+    }
+
+    // check the file can be written
+    FILE* fh = fopen(filepath,"wb");
+    if( !fh )
+    {
+        std::cerr << "write_csv: file [" << filepath << "] could not be opened\n";
+        return GDF_FILE_ERROR;
+    }
+
+    //
+    // It would be good if we could chunk this.
+    // Use the rows*count calculation and a memory threshold to
+    // output a subset of rows at a time instead of the whole thing at once.
+    // The entire CSV must fit in CPU memory before writing it out.
+    //
+    // compute sizes we need
+    // build a matrix of pointers
+    std::unique_ptr<int> pdatalens(new int[rows*count]); // matrix
+    int* datalens = pdatalens.get();
+    size_t memsize = 0;
+    for( unsigned int idx=0; idx < count; ++idx )
+    {
+        gdf_column* col = columns[idx];
+        const char* delim = ((idx+1)<count ? delimiter : terminator);
+        NVStrings* sdata = column_to_strings_csv(col,delim,true_value,false_value);
+        if( sdata )
+            memsize += sdata->byte_count(datalens + (idx*rows),false);
+        if( sdata )
+            NVStrings::destroy(sdata);
+    }
+
+    // cols/rows are transposed in this diagram
+    // datalens
+    //    1,  1,  2, 11, 12,  7,  7 =  41
+    //    1,  1,  2,  2,  3,  7,  6 =  22
+    //   20, 20, 20, 20, 20, 20, 20 = 140
+    //    5,  6,  4,  6,  4,  4,  5 =  34
+    //   --------------------------------
+    //   27, 28, 28, 39, 39, 38, 38 = 237
+    //
+    //
+    // this is a vertical-scan plus carry:
+    // dataptrs
+    //     0,  27,  55,  83, 122, 161, 199
+    //     1,  28,  57,  94, 134, 168, 206
+    //     2,  29,  59,  96, 137, 175, 213
+    //    22,  49,  79, 116, 157, 195, 233
+    //
+    // looks like if we transposed, this would be exclusive-scan but then we have to untranspose;
+    // so it may be better to fixup in place
+    //
+    //printf("memsize=%ld\n",memsize);
+    std::unique_ptr<char> pbuffer(new char[memsize+1]);
+    char* buffer = pbuffer.get();
+    memset(buffer,',',memsize); // fill with commas
```

Comment:
How about a `std::vector<char> buffer(memsize, ',')` instead? 3 lines of code become one and we avoid C-isms like `memset`.

Reply:
The memset was actually not supposed to be there anymore.
Using std::vector for now but may switch to host_vector (or device_vector) in the future.